node/pkg/algorand/watcher.go

```diff
@@ -2,29 +2,250 @@ package algorand
+		debug        bool
 	}
 )
 
-// NewWatcher creates a new Algorand contract watcher
-func NewWatcher(urlRPC string, urlToken string, contract string, lockEvents chan *common.MessagePublication, setEvents chan *common.GuardianSet) *Watcher {
-	return &Watcher{urlRPC: urlRPC, urlToken: urlToken, contract: contract, msgChan: lockEvents, setChan: setEvents}
+var (
+	algorandMessagesConfirmed = promauto.NewCounter(
+		prometheus.CounterOpts{
+			Name: "wormhole_algorand_observations_confirmed_total",
+			Help: "Total number of verified Algorand observations found",
+		})
+	currentAlgorandHeight = promauto.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "wormhole_algorand_current_height",
+			Help: "Current Algorand block height",
+		})
+)
+
+
+// NewWatcher creates a new Algorand appid watcher
+func NewWatcher(
+	indexerRPC   string,
+	indexerToken string,
+	appid        uint64,
+	lockEvents   chan *common.MessagePublication,
+	setEvents    chan *common.GuardianSet,
+	obsvReqC     chan *gossipv1.ObservationRequest,
+) *Watcher {
+	return &Watcher{
+		indexerRPC:   indexerRPC,
+		indexerToken: indexerToken,
+		appid:        appid,
+		msgChan:      lockEvents,
+		setChan:      setEvents,
+		obsvReqC:     obsvReqC,
+		next_round:   0,
+		debug:        true,
+	}
+}
+
+func lookAtTxn(e *Watcher, t models.Transaction, logger *zap.Logger) {
+	if len(t.InnerTxns) > 0 {
+		for q := 0; q < len(t.InnerTxns); q++ {
+			var it = t.InnerTxns[q]
+			var at = it.ApplicationTransaction
+
+			if (len(at.ApplicationArgs) != 3) || (at.ApplicationId != e.appid) || (len(it.Logs) == 0) {
+				continue
+			}
+
+			if string(at.ApplicationArgs[0]) != "publishMessage" {
+				continue
+
+			}
+
+			if e.debug {
+				JSON, _ := json.Marshal(it)
+				logger.Info(string(JSON))
+			}
+
+			emitter, err := types.DecodeAddress(it.Sender)
+			if nil != err {
+				logger.Info(err.Error())
+				continue;
+			}
+
+			var a vaa.Address
+			copy(a[:], emitter[:]) // 32 bytes = 8edf5b0e108c3a1a0a4b704cc89591f2ad8d50df24e991567e640ed720a94be2
+                                                             
+			if e.debug {
+				logger.Info("emitter: " + hex.EncodeToString(emitter[:]))
+			}
+
+			id, err := base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(t.Id)
+			if nil != err {
+				logger.Info(err.Error())
+				continue;
+			}
+
+			if e.debug {
+				logger.Info("id: " + hex.EncodeToString(id) + " " + t.Id)
+			}
+
+			var txHash = eth_common.BytesToHash(id)   // 32 bytes = d3b136a6a182a40554b2fafbc8d12a7a22737c10c81e33b33d1dcb74c532708b
+
+			observation := &common.MessagePublication{
+				TxHash:           txHash,
+				Timestamp:        time.Unix(int64(it.RoundTime), 0),
+				Nonce:            uint32(binary.BigEndian.Uint64(at.ApplicationArgs[2])),
+				Sequence:         binary.BigEndian.Uint64(it.Logs[0]),
+				EmitterChain:     vaa.ChainIDAlgorand,
+				EmitterAddress:   a,
+				Payload:          at.ApplicationArgs[1],
+				ConsistencyLevel: 32,   // What SHOULD this be?
```

Comment:
Per [these docs](https://algorand.foundation/algorand-protocol/core-blockchain-innovation):

```
And Algorand is the first blockchain to provide immediate transaction finality. No forking. No uncertainty.
```

For other instant finality chains such as Terra, [we set the ConsistencyLevel to 0](https://github.com/certusone/wormhole/blob/9ed71c00380ffdbc5230533e85ba58b60c28c4b3/node/pkg/terra/watcher.go#L463)

Reply:
ty